File: src/appl/dqrls.f
c
c     dqrfit is a subroutine to compute least squares solutions
c     to the system
c
c     (1)               x * b = y
c
c     which may be either under-determined or over-determined.
c     the user must supply a tolerance to limit the columns of
c     x used in computing the solution.  in effect, a set of
c     columns with a condition number approximately bounded by
c     1/tol is used, the other components of b being set to zero.
c
c     on entry
c
c        x      double precision(n,p).
c               x contains n-by-p coefficient matrix of
c               the system (1), x is destroyed by dqrfit.
c
c        n      the number of rows of the matrix x.
c
c        p      the number of columns of the matrix x.
c
c        y      double precision(n,ny)
c               y contains the right hand side(s) of the system (1).
c
c        ny     the number of right hand sides of the system (1).
c
c        tol    double precision
c               tol is the nonnegative tolerance used to
c               determine the subset of columns of x included
c               in the solution.  columns are pivoted out of
c               decomposition if
c
c        jpvt   integer(p)
c               the values in jpvt are permuted in the same
c               way as the columns of x.  this can be useful
c               in unscrambling coefficients etc.
c
c        work   double precision(2*p)
c               work is an array used by dqrdc2 and dqrsl.
c
c     on return
c
c        x      contains the output array from dqrdc2.
c               namely the qr decomposition of x stored in
c               compact form.
c
c        b      double precision(p,ny)
c               b contains the solution vectors with rows permuted
c               in the same way as the columns of x.  components
c               corresponding to columns not used are set to zero.
c
c        rsd    double precision(n,ny)
c               rsd contains the residual vectors y-x*b.
c
c        qty    double precision(n,ny)     t
c               qty contains the vectors  q y.   note that
c               the initial p elements of this vector are
c               permuted in the same way as the columns of x.
c
c        k      integer
c               k contains the number of columns used in the
c               solution.
c
c        jpvt   has its contents permuted as described above.
c
c        qraux  double precision(p)
c               qraux contains auxiliary information on the
c               qr decomposition of x.
c
c
c     on return the arrays x, jpvt and qraux contain the
c     usual output from dqrdc, so that the qr decomposition
c     of x with pivoting is fully available to the user.
c     in particular, columns jpvt(1), jpvt(2),...,jpvt(k)
c     were used in the solution, and the condition number
c     associated with those columns is estimated by
c     abs(x(1,1)/x(k,k)).
c
c     dqrfit uses the linpack routines dqrdc and dqrsl.
c
      subroutine dqrls(x,n,p,y,ny,tol,b,rsd,qty,k,jpvt,qraux,work)
      integer n,p,ny,k,jpvt(p)
      double precision x(n,p),y(n,ny),tol,b(p,ny),rsd(n,ny),
     .                 qty(n,ny),qraux(p),work(p)
c
c     internal variables.
c
      integer info,j,jj,kk
c
c     reduce x.
c
      call dqrdc2(x,n,n,p,tol,k,qraux,jpvt,work)
c
c     solve the truncated least squares problem for each rhs.
c
      if(k .gt. 0) then
         do 20 jj=1,ny
   20       call dqrsl(x,n,n,k,qraux,y(1,jj),rsd(1,jj),qty(1,jj),
     1           b(1,jj),rsd(1,jj),rsd(1,jj),1110,info)
      else
         do 30 i=1,n
            do 30 jj=1,ny
   30           rsd(i,jj) = y(i,jj)
      endif
c
c     set the unused components of b to zero.
c
      kk = k + 1
      do 50 j=kk,p
         do 40 jj=1,ny
            b(j,jj) = 0.d0
   40    continue
   50 continue
      return
      end

File: src/appl.c/dqrls.c
/* -----------------------------------------------------------------------
 *
 *  R : A Computer Langage for Statistical Data Analysis
 *  Copyright (C) 1996, 1997  Robert Gentleman and Ross Ihaka
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * -----------------------------------------------------------------------
 *
 *     dqrfit is a subroutine to compute least squares solutions
 *     to the system
 *
 *     (1)               x * b = y
 *
 *     which may be either under-determined or over-determined.
 *     the user must supply a tolerance to limit the columns of
 *     x used in computing the solution.  in effect, a set of
 *     columns with a condition number approximately bounded by
 *     1/tol is used, the other components of b being set to zero.
 *
 *     on entry
 *
 *        x      double precision(n,p).
 *               x contains n-by-p coefficient matrix of
 *               the system (1), x is destroyed by dqrfit.
 *
 *        n      the number of rows of the matrix x.
 *
 *        p      the number of columns of the matrix x.
 *
 *        y      double precision(n,ny)
 *               y contains the right hand side(s) of the system (1).
 *
 *        ny     the number of right hand sides of the system (1).
 *
 *        tol    double precision
 *               tol is the nonnegative tolerance used to
 *               determine the subset of columns of x included
 *               in the solution.  columns are pivoted out of
 *               decomposition if
 *
 *        jpvt   int(p)
 *               the values in jpvt are permuted in the same
 *               way as the columns of x.  this can be useful
 *               in unscrambling coefficients etc.
 *
 *        work   double precision(2*p)
 *               work is an array used by dqrdc2 and dqrsl.
 *
 *     on return
 *
 *        x      contains the output array from dqrdc2.
 *               namely the qr decomposition of x stored in
 *               compact form.
 *
 *        b      double precision(p,ny)
 *               b contains the solution vectors with rows permuted
 *               in the same way as the columns of x.  components
 *               corresponding to columns not used are set to zero.
 *
 *        rsd    double precision(n,ny)
 *               rsd contains the residual vectors y-x*b.
 *
 *        qty    double precision(n,ny)     t
 *               qty contains the vectors  q y.   note that
 *               the initial p elements of this vector are
 *               permuted in the same way as the columns of x.
 *
 *        k      int
 *               k contains the number of columns used in the
 *               solution.
 *
 *        jpvt   has its contents permuted as described above.
 *
 *        qraux  double precision(p)
 *               qraux contains auxiliary information on the
 *               qr decomposition of x.
 *
 *
 *     on return the arrays x, jpvt and qraux contain the
 *     usual output from dqrdc, so that the qr decomposition
 *     of x with pivoting is fully available to the user.
 *     in particular, columns jpvt(1), jpvt(2),...,jpvt(k)
 *     were used in the solution, and the condition number
 *     associated with those columns is estimated by
 *     abs(x(1,1)/x(k,k)).
 *
 *     dqrfit uses the linpack routines dqrdc and dqrsl.
*/

#include "Fortran.h"
#include "Blas.h"
#include "Linpack.h"

static int c__1110 = 1110;

int 
F77_SYMBOL(dqrls) (double *x, int *n, int *p, double
		   *y, int *ny, double *tol, double *b, double *rsd,
		   double *qty, int *k, int *jpvt, double *qraux,
		   double *work)
{
	int x_dim1, x_offset, y_dim1, y_offset, b_dim1, b_offset, rsd_dim1, rsd_offset,
	 qty_dim1, qty_offset, i__1, i__2;
	int info, j;
	extern int F77_SYMBOL(dqrdc2)();
	int jj, kk;

	--work;
	--qraux;
	--jpvt;
	x_dim1 = *n;
	x_offset = x_dim1 + 1;
	x -= x_offset;
	qty_dim1 = *n;
	qty_offset = qty_dim1 + 1;
	qty -= qty_offset;
	rsd_dim1 = *n;
	rsd_offset = rsd_dim1 + 1;
	rsd -= rsd_offset;
	b_dim1 = *p;
	b_offset = b_dim1 + 1;
	b -= b_offset;
	y_dim1 = *n;
	y_offset = y_dim1 + 1;
	y -= y_offset;

	/* reduce x. */

	F77_SYMBOL(dqrdc2) (&x[x_offset], n, n, p, tol, k, &qraux[1], &jpvt[1], &work[1]);

	/* solve the truncated least squares problem for each rhs. */

	if (*k == 0) {
		goto L30;
	}
	i__1 = *ny;
	for (jj = 1; jj <= i__1; ++jj) {
		F77_SYMBOL(dqrsl) (&x[x_offset], n, n, k, &qraux[1], &y[jj * y_dim1 + 1], &rsd[jj
											       * rsd_dim1 + 1], &qty[jj * qty_dim1 + 1], &b[jj * b_dim1 + 1],
		&rsd[jj * rsd_dim1 + 1], &rsd[jj * rsd_dim1 + 1], &c__1110, &
				   info);
	}
L30:

	/* set the unused components of b to zero. */

	kk = *k + 1;
	i__1 = *p;
	for (j = kk; j <= i__1; ++j) {
		i__2 = *ny;
		for (jj = 1; jj <= i__2; ++jj) {
			b[j + jj * b_dim1] = 0.;
		}
	}
	return 0;
}
